### DISCUSSION TITLE
Future of map editor / iD Editor

### DISCUSSION THREAD

**Discussion by gmaclennan** [digidem/mapeo-desktop#558] [Ideas]
## Summary

Should we write our own map editing tool to replace iD Editor in Mapeo?

## Background

We have been using [iD Editor](https://github.com/openstreetmap/iD/) for the UI for editing map features in Mapeo. This has been really great for several reasons:

- The UX for iD is really easy to learn - big, simple buttons for creating points, lines and ways
- The OpenStreetMap data structure works well for us, especially storing data attributes as unstructured tags (e.g. no fixed schema)
- The iD concept of "presets" is really helpful for separating data structure from how data is presented (both on the map and fields for editing)
- iD has a large community around it that has been constantly adding new features

However working with iD in Mapeo has presented some challenges:

1. We need to maintain a fork of iD, and it takes some time to integrate changes/bug fixes/features from iD, and those changes could cause bugs in Mapeo, so we tend to avoid updating and don't get new features and bug fixes
2. iD is not good for editing large features, and features disappear when you zoom out (you can only edit and see data when zoomed in)
3. It is a lot of work to write and maintain any deeper integration between the UI of iD Editor and the rest of Mapeo
4. Performance is a problem when there are complex shapes (e.g. rivers, large non-square areas) on the screen, especially on low-end devices.
5. iD v3 appears to be moving towards an interface more geared towards more advanced users, and appears to sacrifice simplicity for speed of editing for users that understand the tool well.
6. OSM data types (and, by extension, iD) only supports `string` tag values. In Mapeo we support number, boolean and array types (for multi-select fields). Translating between the two adds complexity and potential bugs.
7. iD does not support WYSIWYG editing, e.g. editing a Mapbox-gl styled layer and seeing features as they would appear (icons, line and area styles) on a printed or online map.

(2) and (3) are fairly major limitations. In particular, we need a way to integrate the observation workflow into the map making process, and doing that integration would require a major investment in our own fork of iD. Whilst iD brings a lot of powerful features for working with OSM data, our current user base does not use many of these features. In the end we could spend more time integrating Mapeo into iD and adding features to a fork than we would spend writing our own implementation of map creation / editing tools.

## Moving away from iD

We would loose the features being built by the community around iD, and we would loose some of the drawing tools, but in many cases either (a) we don't need those features or (b) it's high-cost to keep the fork up-to-date and merge new features and bug fixes.

Writing our own map editor from scratch would be hard, but there are some existing libraries we could use:

- [leaflet-geoman](https://geoman.io/leaflet-geoman)
- [mapbox-gl-draw](https://github.com/mapbox/mapbox-gl-draw)
- [nebula.gl](https://nebula.gl/)

nebula.gl is appealing because it already includes most of the drawing modes we would need for Mapeo. It has a dependency on [deck.gl](https://deck.gl/) but the visualizations supported by deck.gl could be really useful for users. It is built on mapbox-gl and supports mapbox map styles.

mapbox-gl-draw is appealing because it is built upon the same mapbox-gl renderer that we already use for map styling. It also has a nice modular structure that allows different [drawing "modes"](https://github.com/mapbox/mapbox-gl-draw/blob/master/docs/MODES.md#available-custom-modes) (drawing tools) to be added as separate modules, e.g. we could add a mode for drawing rivers without any changes to the core library, and we could customize other modes according to our needs. Unfortunately the library does not seem to receive a lot of attention from the Mapbox so the project future is uncertain, but we could always create our own fork of the project - the code it not too hard to understand or too complex, since a lot fo the rendering logic is in mapbox-gl-js (which is well maintained and supported).

Reading through the mapbox-gl-draw API, I think implementing basic editing tools ourselves would not be that complicated. A lot of the complex code in iD Editor is for interacting with the OSM API and storing a local cache of features in the map view. We could side-step a lot of this by:

1. Generate vector tiles from our map + observation in Mapeo Core and display this with mapbox-gl-js
2. Only when a user clicks a feature in the map, load it into the gl-draw layer, and edit it there (we could hide or fade the original feature in the map layer whilst it is in the edit layer).

The sidebar in iD (where it shows details for each feature e.g. all the tags you can edit) we have already replicated for Mapeo Mobile and the observations view of MD, and using our own view for this gives us a lot more flexibility for our users.

We can choose whether to keep the iD concept of "changesets" or just save each edit one at a time as the edit is made.

**Comment by okdistribute** [comment]
This could also be a nice way for someone at mapbox to help... By helping maintain this library and keeping an eye out for issues

**Comment by hackergrrl** [comment]
I'd be excited for a fully WebGL-based editor, maybe building on the
peermaps frontend tech. I've seen what a difference coding the map stuff
closer to the metal makes -- I think we could build something that's way
faster than the web map stuff that's already out there.

**Comment by okdistribute** [comment]
Was recently discussing this with @substack, they're also interested in working on something similar for peermaps.
